marimbabot_audio/scripts/onset_detection.py

```diff
@@ -0,0 +1,547 @@
+        cqt_overlap_num = int(self.window_overlap_t/self.window_all_t*cqt_len)
+        _cqt = cqt[:,cqt_overlap_num:-cqt_overlap_num]
+        _cqt = _cqt[0,:]
+        #db_sequence = librosa.onset.onset_strength(y=self.buffer, sr=self.sr, aggregate=np.max, fmax=self.fmax,n_mels=60)
+#        for each in _cqt.tolist():
+#            self.plotter_pub.publish(each)
+
+        loudness_seq = librosa.power_to_db(cqt**2)
+
+        # rospy.loginfo(f"dB(median:{np.median(db)} max:{np.max(db)}, avg:{np.average(db)}):{db}")
+        onset_env_cqt = librosa.onset.onset_strength(
+            sr=self.sr, S=librosa.amplitude_to_db(cqt, ref=np.max)
+        )
+
+        # detect where the onset happened
+        onsets_cqt_raw = librosa.onset.onset_detect(
+            y=self.buffer,
+            sr=self.sr,
+            hop_length=self.hop_length,
+            onset_envelope=onset_env_cqt,
+            units="time",
+            backtrack=False,
+            # wait= 0.1*self.sr/self.hop_length,
+            delta=4.0,
+            normalize=False,
+        )
+
+        # the whole windows include 0.5 sec overlap at both end, the target windows is only 1 sec.
+        def in_window(o):
+            # make sure the onset inside the target windows
+            return (
+                    o >= self.window_overlap_t and
+                    o < self.window_t + self.window_overlap_t
+            )
+
+        # filter out the onset in the overlap windows
+        onsets_cqt = [
+            o
+            for o in onsets_cqt_raw
+            if in_window(o)
+        ]
+
+        winners_idx = []
+        winner_onsets_times = []
+        durations_len = []
+
+        # publish events and plot visualization
+        for o in onsets_cqt:
+            fundamental_frequency, confidence, winner_idx, winner_pm_idx = self.onset_classification(o)
+
+
+            if winner_idx is not None:
+                # find the y-position of onset in spectrum
+                onset_time = self.onsets_to_time_in_spec(o)
+                # extract the duration and loudness of the onset note
+                duration, duration_len, loudness = self.extract_duration(
+                    onset_time=onset_time,
+                    winner_idx=winner_idx,
+                    cqt=cqt,
+                    loudnes_seq=loudness_seq
+                )
+
+                winners_idx.append(winner_idx)
+                winner_onsets_times.append(onset_time)
+                durations_len.append(duration_len)
+
+
+                # publish the onset note to topic
+                t = self.buffer_time + rospy.Duration(o)
+                no = NoteOnset()
+                no.header.stamp = t
+                note = hz_to_note(fundamental_frequency)
+                no.note = note
+                no.confidence = confidence
+                no.duration = duration
+                no.loudness = loudness
+                self.pub_onset.publish(no)
+
+                rospy.loginfo(
```

Comment:
Consider lowering it to the debug loglovel

Reply:
done